#include "mainwindow.h"
#include "floating_timer.h"
#include "reminder_dialog.h"
#include "ui_mainwindow.h"
#include <QCloseEvent>
#include <QFont>
#include <QInputDialog>
#include <QMessageBox>
#include <QPainter>
#include <QPixmap>
#include <QSoundEffect>

MainWindow::MainWindow(QWidget *parent)
    : QMainWindow(parent), ui(new Ui::MainWindow), timer(new QTimer(this)),
      isWorkPhase(true), completedCycles(0), isDarkTheme(false),
      volume(0.5f), // é»˜è®¤éŸ³é‡50%
      settings(new QSettings("PomodoroApp", "QtPomodoro", this)) {
  ui->setupUi(this);
  floatingTimer = new FloatingTimer(this);

  // ä»è®¾ç½®åŠ è½½é…ç½®
  loadSettings();

  // åˆå§‹åŒ–UI
  setWindowTitle("ç•ªèŒ„æ—¶é’Ÿ");
  ui->timeLabel->setText(
      QString("%1:00").arg(workDuration / 60, 2, 10, QChar('0')));

  // å¦‚æœæœ‰ä¸»é¢˜ï¼Œæ˜¾ç¤ºä¸»é¢˜å†…å®¹ï¼›å¦åˆ™æ˜¾ç¤º"å·¥ä½œé˜¶æ®µ"
  QString displayText =
      currentSessionTheme.isEmpty() ? "å·¥ä½œé˜¶æ®µ" : currentSessionTheme;
  ui->phaseLabel->setText(displayText);

  ui->startButton->setText("å¼€å§‹");
  ui->pauseButton->setText("æš‚åœ");
  ui->resetButton->setText("é‡ç½®");
  ui->cycleLabel->setText(QString("å·²å®Œæˆå‘¨æœŸ: %1").arg(completedCycles));
  ui->themeButton->setText(isDarkTheme ? "æµ…è‰²ä¸»é¢˜" : "æ·±è‰²ä¸»é¢˜");
  ui->autoLockCheckBox->setChecked(enableAutoLock);

  // è¿æ¥ä¿¡å·å’Œæ§½
  connect(timer, &QTimer::timeout, this, &MainWindow::updateTimer);
  connect(ui->startButton, &QPushButton::clicked, this,
          &MainWindow::onStartButtonClicked);
  connect(ui->pauseButton, &QPushButton::clicked, this,
          &MainWindow::onPauseButtonClicked);
  connect(ui->resetButton, &QPushButton::clicked, this,
          &MainWindow::onResetButtonClicked);
  connect(ui->settingsButton, &QPushButton::clicked, this,
          &MainWindow::onSettingsButtonClicked);
  connect(ui->themeButton, &QPushButton::clicked, this,
          &MainWindow::onThemeChanged);
  connect(ui->floatingButton, &QPushButton::clicked, this,
          &MainWindow::toggleFloatingWindow);
  connect(ui->resetPositionButton, &QPushButton::clicked, this,
          &MainWindow::resetFloatingWindowPosition);
  connect(ui->saveThemeButton, &QPushButton::clicked, this, [this]() {
    saveSessionTheme(ui->themeLineEdit->text());
    ui->themeLineEdit->clear();
  });
  connect(ui->exportThemesButton, &QPushButton::clicked, this, [this]() {
    QString defaultFileName =
        QDateTime::currentDateTime().toString("yyyyMMdd-hhmmss") + "_note.md";
    QString filePath =
        QFileDialog::getSaveFileName(this, "å¯¼å‡ºä¸»é¢˜è®°å½•", defaultFileName,
                                     "Markdownæ–‡ä»¶ (*.md);;CSVæ–‡ä»¶ (*.csv)");
    if (!filePath.isEmpty()) {
      QDateTime today = QDateTime::currentDateTime();
      QDateTime weekStart = today.addDays(-7);
      exportSessionThemes(filePath, weekStart, today);
    }
  });
  connect(ui->autoLockCheckBox, &QCheckBox::checkStateChanged, this,
          &MainWindow::onAutoLockChanged);
  connect(ui->volumeSlider, &QSlider::valueChanged, this,
          &MainWindow::onVolumeChanged);

  // åˆå§‹åŒ–æ—¶é—´
  remainingTime = QTime(0, workDuration / 60, workDuration % 60);

  // åˆ›å»ºç³»ç»Ÿæ‰˜ç›˜å›¾æ ‡
  createTrayIcon();

  // æ›´æ–°æ‰˜ç›˜å›¾æ ‡æ˜¾ç¤º
  updateTimer(); // è¿™ä¼šè®¾ç½®åˆå§‹çš„æ‰˜ç›˜å›¾æ ‡

  // åº”ç”¨ä¸»é¢˜
  applyTheme();
}

MainWindow::~MainWindow() {
  saveSettings();
  delete ui;
}

void MainWindow::createTrayIcon() {
  trayIcon = new QSystemTrayIcon(this);

  // åˆ›å»ºä¸€ä¸ªç®€å•çš„å›¾æ ‡ï¼ˆå¦‚æœæ²¡æœ‰ç³»ç»Ÿä¸»é¢˜å›¾æ ‡ï¼‰
  updateTrayIcon(); // ä½¿ç”¨updateTrayIconæ¥ç¡®ä¿å›¾æ ‡æ­£ç¡®åœ°å“åº”ä¸»é¢˜å˜åŒ–

  trayMenu = new QMenu(this);
  QAction *showAction = new QAction("æ˜¾ç¤ºçª—å£", this);
  QAction *hideAction = new QAction("éšè—çª—å£", this);
  QAction *quitAction = new QAction("é€€å‡º", this);

  connect(showAction, &QAction::triggered, this, &MainWindow::showWindow);
  connect(hideAction, &QAction::triggered, this, &MainWindow::hideWindow);
  connect(quitAction, &QAction::triggered, qApp, &QApplication::quit);

  trayMenu->addAction(showAction);
  trayMenu->addAction(hideAction);
  trayMenu->addSeparator();
  trayMenu->addAction(quitAction);

  trayIcon->setContextMenu(trayMenu);
  trayIcon->show();

  connect(trayIcon, &QSystemTrayIcon::activated, this,
          &MainWindow::onTrayIconActivated);
}

void MainWindow::updateTimer() {
  remainingTime = remainingTime.addSecs(-1);
  ui->timeLabel->setText(remainingTime.toString("mm:ss"));

  // æ›´æ–°æµ®åŠ¨çª—å£çš„æ—¶é—´æ˜¾ç¤º
  if (floatingTimer && floatingTimer->isVisible()) {
    floatingTimer->setTime(remainingTime);
  }

  // æ›´æ–°æ‰˜ç›˜å›¾æ ‡æç¤ºå’Œæ ‡é¢˜
  QString tooltip = QString("ç•ªèŒ„æ—¶é’Ÿ - %1: %2åˆ†é’Ÿ")
                        .arg(ui->phaseLabel->text())
                        .arg(remainingTime.minute());
  trayIcon->setToolTip(tooltip);

  // åˆ›å»ºå¸¦è¿›åº¦æ¡å’Œæ—¶é—´æ˜¾ç¤ºçš„å›¾æ ‡
  QPixmap pixmap(250, 80); // æ›´é•¿çš„å›¾æ ‡æä¾›æ›´å¤§æ˜¾ç¤ºç©ºé—´
  pixmap.fill(Qt::transparent);
  QPainter painter(&pixmap);
  painter.setRenderHint(QPainter::Antialiasing);

  // è®¡ç®—è¿›åº¦ç™¾åˆ†æ¯”
  int totalSeconds = remainingTime.minute() * 60 + remainingTime.second();
  int totalDuration = isWorkPhase ? workDuration : breakDuration;
  int progress = 100 - (totalSeconds * 100) / totalDuration;

  // ç»˜åˆ¶å·¦ä¾§åœ†åœˆè¿›åº¦æ¡
  int circleSize = 40;
  int circleX = 2;
  int circleY = (pixmap.height() - circleSize) / 2;

  // ç»˜åˆ¶åœ†åœˆèƒŒæ™¯
  painter.setBrush(isDarkTheme ? Qt::darkGray : Qt::lightGray);
  painter.setPen(Qt::NoPen);
  painter.drawEllipse(circleX, circleY, circleSize, circleSize);

  // ç»˜åˆ¶è¿›åº¦å¼§çº¿
  painter.setPen(QPen(isDarkTheme ? Qt::white : Qt::black, 3));
  int startAngle = 90 * 16;                   // ä»12ç‚¹å¼€å§‹
  int spanAngle = -progress * 360 / 100 * 16; // é¡ºæ—¶é’ˆç»˜åˆ¶
  painter.drawArc(circleX + 2, circleY + 2, circleSize - 4, circleSize - 4,
                  startAngle, spanAngle);

  // åœ¨åœ†åœˆå³ä¾§ç»˜åˆ¶æ—¶é—´æ–‡æœ¬ï¼ˆåªæ˜¾ç¤ºåˆ†é’Ÿæ•°ï¼‰
  painter.setPen(isDarkTheme ? Qt::white : Qt::black);
  QFont font("Arial", 100, QFont::Bold);
  painter.setFont(font);

  QString timeText = QString::number(remainingTime.minute());
  QRect textRect(circleSize + 15, 0, pixmap.width() - circleSize - 15,
                 pixmap.height());
  painter.drawText(textRect, Qt::AlignCenter, timeText);

  trayIcon->setIcon(QIcon(pixmap));

  if (remainingTime == QTime(0, 0)) {
    playSound();
    switchPhase();
  }
}

void MainWindow::onStartButtonClicked() {
  if (!timer->isActive()) {
    timer->start(1000); // 1ç§’è§¦å‘ä¸€æ¬¡
    ui->startButton->setEnabled(false);
    ui->pauseButton->setEnabled(true);
  }
}

void MainWindow::onPauseButtonClicked() {
  if (timer->isActive()) {
    timer->stop();
    ui->startButton->setEnabled(true);
    ui->pauseButton->setText("ç»§ç»­");
  } else {
    timer->start();
    ui->startButton->setEnabled(false);
    ui->pauseButton->setText("æš‚åœ");
  }
}

void MainWindow::onResetButtonClicked() {
  timer->stop();
  isWorkPhase = true;
  workDuration = settings->value("workDuration", 25 * 60).toInt();
  breakDuration = settings->value("breakDuration", 5 * 60).toInt();
  remainingTime = QTime(0, workDuration / 60, workDuration % 60);
  ui->timeLabel->setText(remainingTime.toString("mm:ss"));

  // å¦‚æœæœ‰ä¸»é¢˜ï¼Œæ˜¾ç¤ºä¸»é¢˜å†…å®¹ï¼›å¦åˆ™æ˜¾ç¤º"å·¥ä½œé˜¶æ®µ"
  QString displayText =
      currentSessionTheme.isEmpty() ? "å·¥ä½œé˜¶æ®µ" : currentSessionTheme;
  ui->phaseLabel->setText(displayText);

  ui->startButton->setEnabled(true);
  ui->pauseButton->setEnabled(false);
  ui->pauseButton->setText("æš‚åœ");

  // é‡ç½®åæ›´æ–°æ‰˜ç›˜å›¾æ ‡æ˜¾ç¤º
  updateTimer();
}

void MainWindow::switchPhase() {
  // æ’­æ”¾æ›´æ˜¾è‘—çš„æç¤ºéŸ³
  playSound();
  QTimer::singleShot(500, this, [this]() { playSound(); });
  QTimer::singleShot(1000, this, [this]() { playSound(); });

  if (isWorkPhase) {
    completedCycles++;
    updateCycleCount();
  }

  isWorkPhase = !isWorkPhase;

  if (isWorkPhase) {
    remainingTime = QTime(0, workDuration / 60, workDuration % 60);

    // å¦‚æœæœ‰ä¸»é¢˜ï¼Œæ˜¾ç¤ºä¸»é¢˜å†…å®¹ï¼›å¦åˆ™æ˜¾ç¤º"å·¥ä½œé˜¶æ®µ"
    QString displayText =
        currentSessionTheme.isEmpty() ? "å·¥ä½œé˜¶æ®µ" : currentSessionTheme;
    ui->phaseLabel->setText(displayText);

    // æ˜¾ç¤ºå¼¹çª—æé†’
    ReminderDialog *dialog =
        new ReminderDialog("ä¼‘æ¯ç»“æŸï¼\nå¼€å§‹æ–°çš„ä¸€è½®å·¥ä½œ â°", this);
    dialog->show();

    // ç»§ç»­æ˜¾ç¤ºæ‰˜ç›˜æ¶ˆæ¯
    trayIcon->showMessage("ç•ªèŒ„æ—¶é’Ÿ", "ä¼‘æ¯ç»“æŸï¼Œå¼€å§‹å·¥ä½œï¼",
                          QSystemTrayIcon::Information, 3000);
  } else {
    remainingTime = QTime(0, breakDuration / 60, breakDuration % 60);
    ui->phaseLabel->setText("ä¼‘æ¯é˜¶æ®µ");

    // å¦‚æœå¯ç”¨è‡ªåŠ¨é”å±ï¼Œåˆ™é”å±
    if (enableAutoLock) {
      lockScreen();
    }

    // æ˜¾ç¤ºå¼¹çª—æé†’
    QString message;
    if (enableAutoLock) {
      message =
          QString(
              "å·¥ä½œå®Œæˆï¼\næ­å–œå®Œæˆç¬¬%1ä¸ªç•ªèŒ„é’Ÿ ğŸ‰\nç³»ç»Ÿå·²è‡ªåŠ¨é”å±ï¼Œè¯·ä¼‘æ¯ ğŸŒ¿")
              .arg(completedCycles);
    } else {
      message = QString("å·¥ä½œå®Œæˆï¼\næ­å–œå®Œæˆç¬¬%1ä¸ªç•ªèŒ„é’Ÿ ğŸ‰\nç°åœ¨å¼€å§‹ä¼‘æ¯ ğŸŒ¿")
                    .arg(completedCycles);
    }
    ReminderDialog *dialog = new ReminderDialog(message, this);
    dialog->show();

    // ç»§ç»­æ˜¾ç¤ºæ‰˜ç›˜æ¶ˆæ¯
    trayIcon->showMessage("ç•ªèŒ„æ—¶é’Ÿ",
                          "å·¥ä½œç»“æŸï¼Œå¼€å§‹ä¼‘æ¯ï¼" +
                              QString(enableAutoLock ? " (ç³»ç»Ÿå·²é”å±)" : ""),
                          QSystemTrayIcon::Information, 3000);
  }

  ui->timeLabel->setText(remainingTime.toString("mm:ss"));
  saveSettings();
}

void MainWindow::onSettingsButtonClicked() {
  bool ok;
  int newWorkDuration = QInputDialog::getInt(
      this, "è®¾ç½®å·¥ä½œæ—¶é—´", "å·¥ä½œæ—¶é—´ï¼ˆåˆ†é’Ÿï¼‰:", workDuration / 60, 1, 120, 1,
      &ok);
  if (ok) {
    workDuration = newWorkDuration * 60;
    settings->setValue("workDuration", workDuration); // ç«‹å³ä¿å­˜
  }

  int newBreakDuration = QInputDialog::getInt(
      this, "è®¾ç½®ä¼‘æ¯æ—¶é—´", "ä¼‘æ¯æ—¶é—´ï¼ˆåˆ†é’Ÿï¼‰:", breakDuration / 60, 1, 60, 1,
      &ok);
  if (ok) {
    breakDuration = newBreakDuration * 60;
    settings->setValue("breakDuration", breakDuration); // ç«‹å³ä¿å­˜
  }

  // ç›´æ¥æ›´æ–°æ˜¾ç¤º
  if (isWorkPhase) {
    remainingTime = QTime(0, workDuration / 60, workDuration % 60);
  } else {
    remainingTime = QTime(0, breakDuration / 60, breakDuration % 60);
  }
  ui->timeLabel->setText(remainingTime.toString("mm:ss"));
  updateTimer();
}

void MainWindow::onThemeChanged() {
  isDarkTheme = !isDarkTheme;
  applyTheme();
  saveSettings();
}

void MainWindow::applyTheme() {
  // ä½¿ç”¨QPaletteæ¥è®¾ç½®ä¸»é¢˜ï¼Œé¿å…æ ·å¼è¡¨å½±å“å¸ƒå±€
  QPalette palette;

  if (isDarkTheme) {
    // æ·±è‰²ä¸»é¢˜
    palette.setColor(QPalette::Window, QColor(43, 43, 43));        // ä¸»çª—å£èƒŒæ™¯
    palette.setColor(QPalette::WindowText, QColor(255, 255, 255)); // æ–‡å­—é¢œè‰²
    palette.setColor(QPalette::Base, QColor(64, 64, 64));          // æ§ä»¶èƒŒæ™¯
    palette.setColor(QPalette::AlternateBase, QColor(53, 53, 53)); // äº¤æ›¿èƒŒæ™¯
    palette.setColor(QPalette::ToolTipBase, QColor(255, 255, 255));
    palette.setColor(QPalette::ToolTipText, QColor(43, 43, 43));
    palette.setColor(QPalette::Text, QColor(255, 255, 255));       // æ–‡æœ¬é¢œè‰²
    palette.setColor(QPalette::Button, QColor(64, 64, 64));        // æŒ‰é’®èƒŒæ™¯
    palette.setColor(QPalette::ButtonText, QColor(255, 255, 255)); // æŒ‰é’®æ–‡å­—
    palette.setColor(QPalette::BrightText, QColor(255, 0, 0));
    palette.setColor(QPalette::Link, QColor(173, 216, 230));
    palette.setColor(QPalette::Highlight, QColor(110, 110, 110)); // é€‰ä¸­èƒŒæ™¯
    palette.setColor(QPalette::HighlightedText,
                     QColor(255, 255, 255)); // é€‰ä¸­æ–‡å­—

    ui->themeButton->setText("æµ…è‰²ä¸»é¢˜");
  } else {
    // æµ…è‰²ä¸»é¢˜ï¼ˆç³»ç»Ÿé»˜è®¤ï¼‰
    palette = QApplication::palette(); // æ¢å¤ç³»ç»Ÿé»˜è®¤ä¸»é¢˜
    ui->themeButton->setText("æ·±è‰²ä¸»é¢˜");

    // ç¡®ä¿é‡ç½®æ‰€æœ‰å¿…è¦çš„é¢œè‰²
    palette.setColor(QPalette::Window, QColor(240, 240, 240));
    palette.setColor(QPalette::WindowText, Qt::black);
    palette.setColor(QPalette::Base, Qt::white);
    palette.setColor(QPalette::AlternateBase, QColor(233, 233, 233));
    palette.setColor(QPalette::Text, Qt::black);
    palette.setColor(QPalette::Button, QColor(240, 240, 240));
    palette.setColor(QPalette::ButtonText, Qt::black);
  }

  // åº”ç”¨å…¨å±€è°ƒè‰²æ¿ï¼Œä¿æŒæ§ä»¶å¸ƒå±€ç¨³å®š
  QApplication::setPalette(palette);
  setPalette(palette);

  // æµ®åŠ¨çª—å£ä¹Ÿéœ€è¦åº”ç”¨ä¸»é¢˜ï¼Œä½†ä¸æ”¹å˜å…¶å¸ƒå±€
  if (floatingTimer && floatingTimer->isVisible()) {
    floatingTimer->update();
  }

  // æ›´æ–°æ‰˜ç›˜å›¾æ ‡ä»¥é€‚åº”æ–°ä¸»é¢˜
  updateTrayIcon();
}

void MainWindow::updateCycleCount() {
  ui->cycleLabel->setText(QString("å·²å®Œæˆå‘¨æœŸ: %1").arg(completedCycles));
}

void MainWindow::playSound() {
  QSoundEffect *effect = new QSoundEffect(this);
  effect->setSource(QUrl::fromLocalFile("/System/Library/Sounds/Ping.aiff"));
  effect->setVolume(volume);
  effect->play();
  connect(effect, &QSoundEffect::playingChanged, [effect]() {
    if (!effect->isPlaying()) {
      effect->deleteLater();
    }
  });
}

void MainWindow::onTrayIconActivated(QSystemTrayIcon::ActivationReason reason) {
  if (reason == QSystemTrayIcon::DoubleClick) {
    if (isVisible()) {
      hide();
    } else {
      show();
      activateWindow();
    }
  }
}

void MainWindow::showWindow() {
  show();
  activateWindow();
}

void MainWindow::hideWindow() {
  // åªéšè—ä¸»çª—å£ï¼Œæµ®åŠ¨çª—å£ä¿æŒæ˜¾ç¤ºçŠ¶æ€
  hide();
}

void MainWindow::closeEvent(QCloseEvent *event) {
  if (trayIcon->isVisible()) {
    hide();
    event->ignore();
  }
}

void MainWindow::saveSettings() {
  settings->setValue("workDuration", workDuration);
  settings->setValue("breakDuration", breakDuration);
  settings->setValue("completedCycles", completedCycles);
  settings->setValue("isDarkTheme", isDarkTheme);
  settings->setValue("enableAutoLock", enableAutoLock);
  settings->setValue("volume", volume);
}

void MainWindow::loadSettings() {
  workDuration = settings->value("workDuration", 25 * 60).toInt();
  breakDuration = settings->value("breakDuration", 5 * 60).toInt();
  completedCycles = settings->value("completedCycles", 0).toInt();
  isDarkTheme = settings->value("isDarkTheme", false).toBool();
  enableAutoLock = settings->value("enableAutoLock", false).toBool();
  volume = settings->value("volume", 0.5f).toFloat(); // åŠ è½½éŸ³é‡è®¾ç½®

  // æ›´æ–°éŸ³é‡æ»‘å—å’Œæ˜¾ç¤ºæ ‡ç­¾
  ui->volumeSlider->setValue(static_cast<int>(volume * 100));
  ui->volumeValueLabel->setText(
      QString("%1%").arg(static_cast<int>(volume * 100)));
}

void MainWindow::lockScreen() {
  // è°ƒç”¨ç³»ç»Ÿå‘½ä»¤é”å±ï¼ˆmacOSï¼‰
  system("pmset displaysleepnow");
}

void MainWindow::onAutoLockChanged() {
  enableAutoLock = ui->autoLockCheckBox->isChecked();
  settings->setValue("enableAutoLock", enableAutoLock);
}

void MainWindow::keyPressEvent(QKeyEvent *event) {
  if (event->key() == Qt::Key_Escape) {
    hideWindow();
  }
  QMainWindow::keyPressEvent(event);
}

void MainWindow::toggleFloatingWindow() {
  if (floatingTimer->isVisible()) {
    floatingTimer->hide();
    ui->floatingButton->setText("æµ®åŠ¨çª—å£");
  } else {
    // æ›´æ–°æ—¶é—´åˆ°æµ®åŠ¨çª—å£
    floatingTimer->setTime(remainingTime);
    floatingTimer->setWorkPhase(isWorkPhase);
    floatingTimer->resetTimer(workDuration, breakDuration, isWorkPhase);

    if (timer->isActive()) {
      floatingTimer->startTimer();
    } else {
      floatingTimer->stopTimer();
    }

    floatingTimer->show();
    ui->floatingButton->setText("éšè—æµ®åŠ¨");
  }
}

void MainWindow::resetFloatingWindowPosition() {
  if (floatingTimer) {
    floatingTimer->moveToDefaultPosition();
    floatingTimer->show(); // ç¡®ä¿çª—å£æ˜¾ç¤º
  }
}

// ä¿å­˜å½“å‰ä¼šè¯ä¸»é¢˜
void MainWindow::saveSessionTheme(const QString &theme) {
  currentSessionTheme = theme;
  QSettings settings("PomodoroApp", "SessionThemes");
  settings.beginGroup("Themes");
  settings.setValue(QDateTime::currentDateTime().toString(Qt::ISODate), theme);
  settings.endGroup();

  // ç«‹å³æ›´æ–°ç•Œé¢æ˜¾ç¤º
  QString displayText =
      currentSessionTheme.isEmpty() ? "å·¥ä½œé˜¶æ®µ" : currentSessionTheme;
  ui->phaseLabel->setText(displayText);
}

// è·å–æŒ‡å®šæ—¶é—´èŒƒå›´å†…çš„ä¸»é¢˜è®°å½•
QList<QPair<QDateTime, QString>>
MainWindow::getSessionThemes(const QDateTime &from, const QDateTime &to) const {
  QList<QPair<QDateTime, QString>> result;
  QSettings settings("PomodoroApp", "SessionThemes");
  settings.beginGroup("Themes");

  foreach (const QString &key, settings.allKeys()) {
    QDateTime dt = QDateTime::fromString(key, Qt::ISODate);
    if (dt >= from && dt <= to) {
      result.append(qMakePair(dt, settings.value(key).toString()));
    }
  }

  settings.endGroup();
  return result;
}

// å¯¼å‡ºä¸»é¢˜è®°å½•åˆ°æ–‡ä»¶
void MainWindow::exportSessionThemes(const QString &filePath,
                                     const QDateTime &from,
                                     const QDateTime &to) const {
  QFile file(filePath);
  if (!file.open(QIODevice::WriteOnly | QIODevice::Text)) {
    return;
  }

  QTextStream out(&file);
  out << "æ—¶é—´,ä¸»é¢˜å†…å®¹\n";

  auto themes = getSessionThemes(from, to);
  foreach (const auto &theme, themes) {
    out << theme.first.toString("yyyy-MM-dd hh:mm:ss") << "," << theme.second
        << "\n";
  }

  file.close();
}

// éŸ³é‡è°ƒèŠ‚æ§½å‡½æ•°
void MainWindow::onVolumeChanged(int value) {
  volume = value / 100.0f;              // å°†0-100çš„å€¼è½¬æ¢ä¸º0.0-1.0
  settings->setValue("volume", volume); // ä¿å­˜éŸ³é‡è®¾ç½®

  // æ›´æ–°éŸ³é‡æ˜¾ç¤ºæ ‡ç­¾
  ui->volumeValueLabel->setText(QString("%1%").arg(value));
}

// æ›´æ–°æ‰˜ç›˜å›¾æ ‡ä»¥é€‚é…å½“å‰ä¸»é¢˜
void MainWindow::updateTrayIcon() {
  if (!trayIcon)
    return;

  // åˆ›å»ºç®€å•çš„å•è‰²å›¾æ ‡é€‚é…ä¸»é¢˜
  QPixmap pixmap(16, 16);
  pixmap.fill(isDarkTheme ? Qt::white : Qt::black);
  trayIcon->setIcon(QIcon(pixmap));
}